File: src/main/java/com/now/nowbot/model/beatmapParse/parse/ManiaBeatmapAttributes.java
package com.now.nowbot.model.beatmapParse.parse;

import com.now.nowbot.model.beatmapParse.HitObject;

import java.io.BufferedReader;
import java.io.IOException;

public class ManiaBeatmapAttributes extends OsuBeatmapAttributes {

    public ManiaBeatmapAttributes(BufferedReader reader, BeatmapGeneral general) throws IOException {
        super(reader, general);
        for (HitObject line : hitObjects) {
            var column = getColumn(line.getPosition().getX(), (int) Math.floor(CS));
            line.setColumn(column);
        }
    }

    private static int getColumn(double x, int key) {
        int column = (int) Math.floor(x * key / 512f);
        column = Math.min(Math.max(0, column), key - 1);
        return column;
    }
}

File: src/main/java/com/now/nowbot/model/beatmapParse/parse/ManiaBeatmapAttributes.kt
package com.now.nowbot.model.beatmapParse.parse

import java.io.BufferedReader

class ManiaBeatmapAttributes(reader: BufferedReader, general: BeatmapGeneral?) : OsuBeatmapAttributes(reader, general) {
    init {
        val cs = cs.toInt()

        for (line in hitObjects) {
            line.column = getColumn(
                line.position.x, cs
            )
        }
    }

    companion object {
        private fun getColumn(x: Int, key: Int): Int {
            if (key <= 0) return 0
            
            return (x * key / 512.0).toInt().coerceIn(0 ..< key)
        }
    }
}
